### PR TITLE
Initialize Apprentice for generated clients

### PR DESCRIPTION
**Pull request by edwinzg**
As part of the shadowing pipeline, we want to call apprentice instead of downstream clients for shadow requests.

This diff initializes Apprentice for tchannel & http generated clients. Will add the code to call Apprentice instead of the actual downstream in a followup diff

### CODE REVIEW

File: codegen/package.go [hunk 1]
```diff
@@ -146,6 +150,13 @@ func (p *PackageHelperOptions) deputyReqHeader() string {
 	return "x-deputy-forwarded"
 }
 
+func (p *PackageHelperOptions) shadowReqHeader() string {
+	if p.ShadowReqHeader != "" {
+		return p.ShadowReqHeader
+	}
+	return "X-Shadow-Start-Trace-Id"
```
[the remark this answers is not in the document]
Comment (cl1337) [reply codegen/package.go]: usually this should be configurable

File: codegen/templates/http_client.tmpl [hunk 2]
```diff
@@ -76,6 +79,37 @@ func {{$exportName}}(deps *module.Dependencies) Client {
 		deps.Default.Config.MustGetStruct("clients.{{$clientID}}.defaultHeaders", &defaultHeaders)
 	}
 
+	{{if ne $clientID "apprentice" -}}
+	apprenticeMethodNames := map[string]string{
+		"Apprentice::getRequest": "getRequest",
+		"Apprentice::getResponse": "getResponse",
+	}
+
+	apprenticeTimeout := time.Millisecond * time.Duration(
```
Comment (jacobgreenleaf) [inline codegen/templates/http_client.tmpl]: Can you add an entry to the changelog that documents this? Now `clients.apprentice.timeout` and others are required. Or you can make them optional.

Comment (edwinzg) [reply codegen/templates/http_client.tmpl]: i'll make it optional

**Comment by edwinzg** [comment]
@cl1337 what is stagingReqHeaders/how is it used?

File: codegen/package.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (edwinzg) [reply codegen/package.go]: I copied what other things did - I think people can pass in ShadowReqHeader for their own needs, this is just the default

**Comment by cl1337** [comment]
@edwinzg if `stagingReqHeader` presents, client will try to use another channel to route request to a different service: https://github.com/uber/zanzibar/blob/master/examples/example-gateway/build/clients/baz/baz.go#L283

I think this was designed to route traffic to apprentice at the first place

**Pull request merged**